### PR TITLE
Update config schema to latest spec

### PR DESCRIPTION
**Pull request by ivange94**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -59,7 +59,7 @@
     "@babel/preset-react": "^7.0.0",
     "@babel/preset-typescript": "^7.3.3",
     "@openmrs/esm-api": "^3.0.8",
```
Comment (FlorianRappl) [inline package.json:63]: Great - I'd just update all the (core) dependencies to their latest version (3.1.1).

**Review by FlorianRappl — COMMENTED** [review]
I think this is great!

**Comment by ivange94** [comment]
@FlorianRappl any idea why I'm getting the `TypeError`? I've updated all core dependencies locally but I still get that error.

File: src/config-schemas/openmrs-esm-patient-registration-schema.ts [hunk 2]
```diff
@@ -29,16 +47,24 @@ export const esmPatientRegistrationSchema = {
...
-      default: '${openmrsSpaBase}/patient/${patientUuid}/chart',
-      validators: [validators.isUrlWithTemplateParameters(['patientUuid'])],
+    _type: Type.Object,
+    _default: {
+      _elements: {
+        submitButton: {
+          _type: Type.String,
+          _default: '${openmrsSpaBase}/patient/${patientUuid}/chart',
+        },
+      },
```
Comment (brandones) [inline src/config-schemas/openmrs-esm-patient-registration-schema.ts]: You should never have schema keywords nested in `_default`. `_default` should just be, literally, the default value. And use of `Type.Object` should be very rare.

```suggestion
    submitButton: {
      _type: Type.String,
      _default: '${openmrsSpaBase}/patient/${patientUuid}/chart'
    },
```

File: src/config-schemas/openmrs-esm-patient-registration-schema.ts [hunk 3]
```diff
@@ -1,26 +1,44 @@
-import { validators } from '@openmrs/esm-config';
+import { Type } from '@openmrs/esm-config';
 
 export const esmPatientRegistrationSchema = {
   personAttributeSections: {
-    arrayElements: {
-      name: { default: false, validators: [validators.isString] },
+    _type: Type.Array,
+    _elements: {
```
Comment (brandones) [inline src/config-schemas/openmrs-esm-patient-registration-schema.ts:6]: The objects in `personAttribueSections` should have a top-level key `name`, alongside `personAttribues`. That seems to have been lost.

File: src/config-schemas/openmrs-esm-patient-registration-schema.ts [hunk 4]
```diff
@@ -1,26 +1,44 @@
...
+            _elements: {
+              required: {
+                _type: Type.Boolean,
+                _default: false,
+              },
+              matches: {
+                _type: Type.String,
+                _default: null,
+              },
+            },
```
Comment (brandones) [inline src/config-schemas/openmrs-esm-patient-registration-schema.ts]:
```suggestion
            required: {
              _type: Type.Boolean,
              _default: false,
            },
            matches: {
              _type: Type.String,
              _default: null,
            },
```

File: src/config-schemas/openmrs-esm-patient-registration-schema.ts (hunk 3 above)
Comment (brandones) [reply src/config-schemas/openmrs-esm-patient-registration-schema.ts:6]:
```suggestion
    _elements: {
      name: {
        _type: Type.String,
        _default: false,
      },
```

**Review by brandones — CHANGES_REQUESTED** [review]
Changes needed, as described above.

Sorry I don't have an answer about the typescript failure. I don't know why that's happening. @FlorianRappl ?

**Comment by FlorianRappl** [comment]
I'm not sure what error you get. Sometimes removing the package-lock and node_modules and reinstalling helps.

**Comment by ivange94** [comment]
> I'm not sure what error you get. Sometimes removing the package-lock and node_modules and reinstalling helps.

If you look at the build failure, you'll see that it complains that `TypeError: Cannot read property 'Array' of undefined` when i do `Type.Array` but on the UI this works fine though. The error is only within the tests.

File: src/config-schemas/openmrs-esm-patient-registration-schema.ts (hunk 2 above)
Comment (ivange94) [reply src/config-schemas/openmrs-esm-patient-registration-schema.ts]: I tried this

```typescript
links: {
    _type: Type.Object,
    _elements: {
         _submitButton: {
             _type: Type.String
         }
     },
     _default: {
         submitButton: '${openmrsSpaBase}/patient/${patientUuid}/chart'
     }
}
```

In the browser there's an error that says the schema is bad that it lacks a default. It only worked when I when with the current approach. What is the way forward for this?

**Comment by FlorianRappl** [comment]
Thanks for clarifying @ivange94 . The error comes because of:

```ts
import { Type } from '@openmrs/esm-config';
``` 

remember that for the tests `@openmrs/esm-config` is mocked. The mock does not expose a `Type`. Thus this is `undefined`.

Here is the mock used:

https://github.com/openmrs/openmrs-esm-patient-registration/blob/master/__mocks__/openmrs-esm-module-config.mock.tsx

File: src/config-schemas/openmrs-esm-patient-registration-schema.ts (hunk 2 above)
Comment (brandones) [reply src/config-schemas/openmrs-esm-patient-registration-schema.ts]: Just use the code I suggested above?

Comment (ivange94) [reply src/config-schemas/openmrs-esm-patient-registration-schema.ts]: I didn't understand it. I thought you just commented on my code, it's now I'm seeing that it's a code suggestion. Did you remove the top level `links` element?

File: src/config-schemas/openmrs-esm-patient-registration-schema.ts [hunk 5]
```diff
@@ -29,16 +48,20 @@ export const esmPatientRegistrationSchema = {
...
         ],
       },
     ],
   },
   links: {
     submitButton: {
-      default: '${openmrsSpaBase}/patient/${patientUuid}/chart',
-      validators: [validators.isUrlWithTemplateParameters(['patientUuid'])],
+      _type: Type.String,
+      _default: '${openmrsSpaBase}/patient/${patientUuid}/chart'
```
Comment (brandones) [inline src/config-schemas/openmrs-esm-patient-registration-schema.ts]:
```suggestion
      _default: '${openmrsSpaBase}/patient/${patientUuid}/chart',
      _validators: [validators.isUrlWithTemplateParameters(['patientUuid'])],
```

**Comment by brandones** [comment]
@ivange94 I merged the changes so now it's easier to tell what's going on by reading the code. Is it clear?

**Review by brandones — APPROVED** [review]

**Comment by ivange94** [comment]
Thanks @FlorianRappl and @brandones I've fixed the merge conflicts and also the broken tests.

**Comment by brandones** [comment]
Excellent! Thanks @ivange94 !

**Pull request merged**